src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs

```diff
@@ -680,32 +682,117 @@ public bool Equals([NotNullWhen(true)] string? value, StringComparison compariso
+                // but we need to be careful here with inliner's budget
             }
 
-            if (RuntimeHelpers.IsKnownConstant(b) && b != null && b.Length == 0)
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            static bool EqualsUnrolled_0_to_8(string? a, string b)
             {
-                return a != null && a.Length == 0;
-            }
+                if (a != null)
+                {
+                    if (b.Length == 0)
+                    {
+                        // Fold Equals(a, "") to 'a != null && a.Length == 0'
+                        return a.Length == 0;
+                    }
+                    if (b.Length == 1)
+                    {
+                        // Load Length, ch1, \0 into ulong (so we can skip the Length check)
+                        return ReadUInt64(a, -2) == (((ulong)b[0] << 32) | 1UL);
+                    }
+                    if (b.Length == 2)
+                    {
+                        // Load Length, ch1, ch2 into ulong (so we can skip the Length check)
+                        return ReadUInt64(a, -2) == (((ulong)b[1] << 48) | ((ulong)b[0] << 32) | 2UL);
+                    }
+                    if (b.Length == 3)
+                    {
+                        // Load ch1, ch2, ch3 and \0 into ulong
+                        return a.Length == 3 && ReadUInt64(a) == (((ulong)b[2] << 32) | ((ulong)b[1] << 16) | b[0]);
+                    }
 
-            if (object.ReferenceEquals(a, b))
-            {
-                return true;
+                    ulong v1 = ReadUInt64(a);
+                    ulong cns1 = ((ulong)b[3] << 48) | ((ulong)b[2] << 32) | ((ulong)b[1] << 16) | b[0];
+                    if (b.Length == 4)
+                    {
+                        // Load ch1, ch2, ch3 and ch4 into ulong
+                        return a.Length == 4 && v1 == cns1;
+                    }
+                    // Handle Length [5..8] via two ulong (overlapped)
+                    return v1 == cns1 && ReadUInt64(a, a.Length - 4) ==
+                        (((ulong)b[b.Length - 1] << 48) | ((ulong)b[b.Length - 2] << 32) | ((ulong)b[b.Length - 3] << 16) | b[b.Length - 4]);
+                }
+
+                // a is null when b is a known non-null
+                return false;
             }
 
-            if (a is null || b is null || a.Length != b.Length)
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            static bool EqualsUnrolled_9_to_16(string? a, string b)
             {
+                if (a != null)
+                {
+                    // Load 'a' into two vectors with overlapping.
+                    Vector128<ushort> v2 = Vector128.LoadUnsafe(ref Unsafe.As<char, ushort>(ref a._firstChar), (nuint)a.Length - 8);
+                    Vector128<ushort> v1 = Vector128.LoadUnsafe(ref Unsafe.As<char, ushort>(ref a._firstChar));
```

Comment:
This may be accessing memory after the end of the string. How do you that it is safe and it won't AV?

Reply:
That's the trick - `v2` loads part of the `v1`'s data so they both never go out of bounds, e.g. this snippet showcases it https://gist.github.com/EgorBo/c8e8490ddd6f9a0d5b72c413ddd81d44#file-avx-string-eq-inlined-cs-L50
PS: In this code v2 and v1 reordered for slightly better codegen.

Reply:
I understand the trick with overlapping data. It is not my concern.

The case that I am concerned about is a.Length == 1. I do not see anything that guarantees there is Vector128 worth of data.

Reply:
Oops, yes, totally forgot about that, added, thanks!